### ISSUE TITLE
Character limit content review for sitewide VA.gov

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#60351]
## About your team

[This was originally caught within the 526ez form by the benefits portfolio team](https://github.com/department-of-veterans-affairs/va.gov-team/issues/42476#issuecomment-1593046794). 

Notably the following engineers have given this alot of thought:

- @tblackwe 
- @kylesoskin 
- @Mottie 

I've raised awareness with the 526ez designers as well
- @lydiahooper 
- @Nyanteea 

This is a sitewide problem that occurs for _any_ input or text area with character limits.

 
### What does your team need support for? Check all that apply.
- [ ] Launching one or more unauthenticated pages, online forms, apps, or tools on VA.gov
- [x] Making revisions to one or more existing VA.gov pages
- [ ] Accessibility review of wireframes and/or prototypes
- [ ] Accessibility issue (general request)
- [ ] Accessibility PDF review request
- [ ] Accessibility user research support ([submit a separate, additional request](https://github.com/department-of-veterans-affairs/va.gov-team/issues/new?assignees=coforma-terry%2Cangelafowler82&labels=508%2FAccessibility%2C+sitewide+CAIA%2C+sitewide+accessibility%2C+CAIA-a11y-research&template=caia-a11y-research.yaml&title=%5BCAIA+A11y+Research%5D%3A+Team+name%2C+product+name))
- [ ] Something else 
 
### Give a brief description of what your team needs support with:

We need review from the sitewide content team on how to resolve a bug related to character limits. To summarize the problem:

##### In the 526ez flow, there’s inputs with character limits like a text area where Veterans can provide supporting evidence.
- Take an input that has a 500 character limit
- Multibyte characters in supporting evidence uploads is a problem
- It will appear as though they are within the 500 character limit but
  - When submitted to the backend, it will go over that character limit and get rejected
  - e.g. Smart-quote (“) would take up more characters than a regular ascii quote (“)

##### How can we fix this? We’ve thought of:
- Adjusting the submission in the backend e.g. automatically swapping the smart quote with a regular ascii quote after the user submits
- Adjusting the submission in the frontend e.g. as the user types, the character is replaced
- Providing error validation in the frontend e.g. when the user attempts to submit, they’re told why they can’t use X characters and asked to replace them (or maybe we could provide a method to automatically fix it, but that would involve more work I’m guessing)

##### Why this gets tricky
- Automatically adjusting submissions could lead to language related changes e.g. ÿ -> y or ñ -> n
- If we update a submission in the backend, users will not know this change has been made
- If we update a submission in the frontend
  - If done live as they type, screen reader users will not be alerted of the change (unless we put together a really tedious contraption to do so)
  - If we provide error validation explaining why we can’t accept certain characters, it may create a ton of friction to go back and fix all the issues
- Even if we fixed it a little more upstream in Lighthouse, the multi-byte issue will persist since the upstream service (BGS) only accepts ASCII characters

##### Possible questions
- If we were to red team this or think of all the worst possible unintended outcomes, what are we looking at? e.g.
  - could small character changes within a name or details in supporting evidence hurt a Veteran’s claim or cause confusion?
  - could small character changes to language e.g. ñ -> n give the impression of an English-speaking bias to Veterans who english isn’t their first language?
- Is this a binary between correcting all characters or none? Or can we write in exceptions that may be more sensitive?

I'm curious to get @DanielleThierryUSDSVA 's thoughts on this especially given her experience with inclusive content writing.
 
 ## Timeframe

 ### Will this new product be released incrementally (for instance 25% of users initially)? 
  - [ ] Yes
  - [x] No
 
**Note:** If you checked yes, we'll reach out to discuss details about the content in the react widget.
  
### When do you expect to launch your product to 100% of users?
  
*Please provide an estimated date so our team can create other relevant tickets.*
 
### Will this work be going through the Collaboration Cycle?
- [ ] Yes 
- [x] No
 
#### If yes, please check all of the phases in the Collaboration Cycle you have completed: 
- [ ] Design Intent
- [ ] Midpoint Review
- [ ] Staging
 
### When does this work need to be completed?
- Estimated launch date: n/a
- Estimated staging review date: n/a 
- Content, Accessibility, and IA work needed by: ASAP
  
## Supporting artifacts
*Please provide supporting artifacts as available.*

[This ticket outlines the issue in more detail](https://github.com/department-of-veterans-affairs/va.gov-team/issues/42476#issuecomment-1593046794)

#### Do you plan to bring this to an upcoming content office hours session?
- [ ] Yes, benefit content office hours (Thursdays, 3:00 p.m. to 3:45 p.m. ET)
- [ ] Yes, health  content office hours (Thursdays, 11:00 a.m. to 11:30 a.m. ET)
- [ ] Yes, unauth office hours (Mondays, 11:00 a.m. to 11:30 a.m. ET)
- [x] No, but I'd like to schedule time to talk about this request
- [ ] No, let's work asynchronously and meet if needed

 **Note:** If we think this work would benefit from a collaborative session with you, we may ask you to bring it to office hours or set up a separate time to meet.
 
### Accessibility Help in Slack
The [#accessibility-help channel](https://dsva.slack.com/archives/C8E985R32) in the DSVA Slack, is also available to you as a resource. Everyone is welcome to ask questions or get help from our a11y (accessibility) specialists. Tag `@Terry Nichols` to get a11y help asap.

## Next steps

**Once you’ve submitted this ticket, please post a link to this issue in the [#sitewide-content-ia](https://dsva.slack.com/channels/sitewide-content-ia) Slack channel and tag Randi Hecht.**

- If you also need engineering support from the Public Websites team, fill out their [intake request form](https://github.com/department-of-veterans-affairs/va.gov-team/issues/new?assignees=Public+Websites%2C+Sitewide+content&labels=vsa-public-websites%2C+vsa%2C+vsa-public-websites-intake%2C+sitewide-content%2C+needs-grooming&template=public-websites-intake.md&title=%3CType+of+Request%3E+from+%3CTeam%3E).
  
- If you need a page/URL redirected, a URL changed or a vanity URL set up, please submit a [Redirect, URL change, or vanity URL request](https://github.com/department-of-veterans-affairs/va.gov-team/issues/new?assignees=mnorthuis&labels=ia&template=redirect-request.md&title=Redirect+Request)
 
 - If you are creating an [experimental design,](https://design.va.gov/about/contributing-to-the-design-system/experimental-components-and-patterns#what-is-an-experimental-component-or-pattern?) please [also contact the DST. ](https://design.va.gov/about/contributing-to-the-design-system/suggest-an-addition-or-update)
- If you are requesting [a11y support for user research,](https://github.com/department-of-veterans-affairs/va.gov-team/issues/new?assignees=coforma-terry%2Cangelafowler82&labels=508%2FAccessibility%2C+sitewide+CAIA%2C+sitewide+accessibility%2C+CAIA-a11y-research&template=caia-a11y-research.yaml&title=%5BCAIA+A11y+Research%5D%3A+Team+name%2C+product+name) please complete this form, next.

**Comment by joshkimux** [comment]
I followed up with @DanielleThierryUSDSVA on this, who drafted some recommendations for how we can start working on this now and later ❤️ 

### Now:
- Conduct an inventory audit on all variations of the 500 character count text area
  - We want to determine if they currently have hint text set up to see if the following solution can work (or fit in)
- If we determine this is feasible, we will attempt to do the following:
  - Automatically replace characters in the front-end
  - Provide hint text programmatically linked to the text area and an additional info component explaining character limitations (Danielle wrote a bunch of copy to test out)
  - This would address the concerns around editing without screen reader users knowing by providing them a heads up before they begin work within the text area.

### Next:
- Danielle is going to start working/investigating more upstream to see how we can tackle this at the most upstream service upstream service (BGS?)
- This will have urgency as it’s a challenge they likely want to fix before having Spanish language forms

### Later:
- Fix it at its most upstream root so we can remove bandaid solutions in our now recommendations

**Comment by sara-amanda** [comment]
@coforma-terry moving this in Projects to match ZenHub in New Intake Tickets. Also, assigning to @ohgreatkate since it is related to Benefits.

**Comment by coforma-jamie** [comment]
@joshkimux, was just assigned to this ticket and wanted to check in. 
- Has the inventory audit for the 500 character text area variations been started yet?  
- I was also wondering if there's a list of exactly which characters are causing the issue, and _how many_ characters each is taking up vs its equivalent (is a 'regular-quote' 1 character and a 'smart-quote' 2, for example)?

Let me know if there's anything I can do to help overall! :)

**Comment by ohgreatkate** [comment]
@tblackwe has your team been gathering an inventory audit? If so how are you logging this effort and how can we support?

> Conduct an inventory audit on all variations of the 500 character count text area
> We want to determine if they currently have hint text set up to see if the following solution can work (or fit in)

**Comment by DanielleThierryUSDSVA** [comment]
@joshkimux @coforma-jamie FYSA - I met with the 10-10CG team to surface this issue in the context of a Spanish-language form experience. Josh, this is the team I mentioned would be the first team to translate an actual form. They're going to look into this further as well.

cc @batemapf @hdjustice @davidakennedy from 10-10CG

**Comment by joshkimux** [comment]
@coforma-jamie excellent questions

> Has the inventory audit for the 500 character text area variations been started yet?

It has not unfortunately, and this task still needs an owner with bandwidth to take on 💙 
Seeing @DanielleThierryUSDSVA 's recent updates here, this might be something the 10-10 team @davidakennedy could look into?

> I was also wondering if there's a list of exactly which characters are causing the issue, and how many characters each is taking up vs its equivalent (is a 'regular-quote' 1 character and a 'smart-quote' 2, for example)?

Also uncertain, I don't have a concrete list-- but we could start by reaching out to some engineers threaded here for help.

@DanielleThierryUSDSVA 

> We'll present information to you in your language, but you can't reliably use the characters of your language.

Biggest of oophs here. This is rough.

**Comment by joshkimux** [comment]
@coforma-jamie we considered this, but were concerned about who would be reading the outputs. We don't have (from my knowledge) a service level viewpoint/understanding on the consequences this may have in the same way your previous team was familiar with. If we could get a better understanding on how these submission are used and the possible consequences (if any), this may be a possible short-term solution to consider 😄

**Comment by DanielleThierryUSDSVA** [comment]
@coforma-jamie @joshkimux I think for the Spanish-language 10-10CG at least, we could ask the team to ask how the forms will be reviewed and how we might communicate this to the folks reviewing Spanish-language submissions.

**Comment by coforma-jamie** [comment]
That's a great idea @DanielleThierryUSDSVA. If we understand how the staff processes a form, it might give us more options (or at least, help us better understand what our options are.)

**Comment by tblackwe** [comment]
> @tblackwe, @kylesoskin, @Mottie: Do you have a list of exactly WHICH characters have a value of WHAT? ("A smart quote is counted as 2 characters," for example)

I don't have a list, and I'm pretty sure it is untenably large. @kylesoskin/ @Mottie any thoughts? we can always drop an array of likely characters in a form and see what comes back.

**Comment by coforma-jamie** [comment]
My thought was, could we possibly hit some low-hanging fruit? Replacing a smart quote with a tick quote won't really impact things, compared to something like Spanish-language characters

**Comment by hdjustice** [comment]
My team is meeting with our CG stakeholders today, and our main topic is to discuss how CARMA manages non-english characters, specifically those with accents or other additional marks.  I will ask about the review process of non-english language application processing as well.

Also FYI - I did get confirmation on how the Enrollment System handles non-english characters from Joshua Faulkner.  Perhaps this may help with the 526ez:

> You're talking about database constraints;
> ES will support non english characters but they take up more than one byte in a varchar column, which is why you don't set the constraints to be the same as the input field length constraints if your database is using ISO charset. i.e. last name field can be 30 chars including extended character sets but the DB column is 100 bytes to allow for conversions.
> Haven't had any issues with it aside from some foreign addresses that were in German, couple years ago we had to extend the column to accommodate some of those.

**Comment by Mottie** [comment]
It all depends on what the upstream service will accept/expect. I _think_ BGS only accepts ASCII characters, which is a [very limited set of characters](https://en.wikipedia.org/wiki/ASCII#Character_set) (94 total). If they support [extended ASCII characters (latin supplement)](https://en.wikipedia.org/wiki/List_of_Unicode_characters#Latin-1_Supplement), then we could include accented characters. Anything with a decimal value > 255 gets us into [unicode](https://en.wikipedia.org/wiki/List_of_Unicode_characters#Latin_Extended-A) and mutliple-byte characters.

**Comment by hdjustice** [comment]
Update for 10-10CG discussion.  Today, they process the Spanish paper form and their system removes the accent, but keeps the character.  It is done this way because the Veteran is matched with data in MPI to confirm they are enrolled in VA health care.  They will look into how MPI handles additional characters, but they don't think it does, and don't think they can validate the Veteran with it against MPI.

**Comment by coforma-terry** [comment]
Closing this for now - can reopen if required.
